implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -52,185 +52,226 @@
+        // so the public key will have to be created per request to ensure the rotation (if any) is effective.
+        String kid = getKid(jws);
+        HttpsJwks httpsJwks = httpsJwksSupplier.get();
+        if (httpsJwks != null && authContextInfo.getTokenKeyId() == null)
+            return getHttpsJwk(httpsJwks, kid);
+
+        if (verificationKey == null) {
+            synchronized (this) {
+                if (verificationKey == null) {
+                    String content = contentSupplier.get();
+                    if (content == null)
+                        throw new UnresolvableKeyException(FAILED_TO_READ_LOCATION_MESSAGE);
+
+                    PublicKey localKey = tryAsLocalJwk(content, kid);
+
+                    if (localKey == null) {
+                        localKey = tryAsPEM(content);
+                    }
+                    if (localKey == null) {
+                        throw new UnresolvableKeyException(FAILED_TO_READ_LOCATION_MESSAGE
+                                + authContextInfo.getPublicKeyLocation());
+                    }
+                    verificationKey = localKey;
+                }
+            }
         }
-        return key;
+        return verificationKey;
     }
 
-    private PublicKey tryAsPEM() {
+    private PublicKey tryAsPEM(String content) {
         try {
-            verificationKey = KeyUtils.decodePublicKey(content);
+            return KeyUtils.decodePublicKey(content);
         } catch (Exception e) {
-            log.debug("Failed to read location as PEM", e);
+            LOGGER.debug("Failed to read location as PEM", e);
         }
-        return verificationKey;
+        return null;
     }
 
-    private PublicKey tryAsJWK(JsonWebSignature jws) throws UnresolvableKeyException {
-        String kid = jws.getHeaders().getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER);
-        if (kid != null) {
-            if (authContextInfo.getTokenKeyId() != null && !kid.equals(authContextInfo.getTokenKeyId())) {
-                log.debugf("Invalid token 'kid' header: %s, expected: %s", kid, authContextInfo.getTokenKeyId());
-                throw new UnresolvableKeyException("Invalid token 'kid' header");
+    PublicKey getHttpsJwk(HttpsJwks httpsJwks, String kid) {
+        try {
+            List<JsonWebKey> keys = httpsJwks.getJsonWebKeys();
+            if (kid != null) {
+                for (JsonWebKey currentJwk : keys) {
+                    if (kid.equals(currentJwk.getKeyId())) {
+                        return PublicJsonWebKey.class.cast(currentJwk).getPublicKey();
+                    }
+                }
+            } else if (keys.size() == 1) {
+                return PublicJsonWebKey.class.cast(keys.get(0)).getPublicKey();
             }
-        } else {
-            kid = authContextInfo.getTokenKeyId();
+        } catch (Exception e) {
+            LOGGER.debug("Failed to read location as JWK(S)", e);
```

Comment:
Given we are creating the dedicated functions for different locations, please update to ""Failed to read JWK set from the HTTPS location", something like that.